pkg/requests/builder.go

```diff
@@ -0,0 +1,173 @@
+package requests
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/bitly/go-simplejson"
+)
+
+// Builder allows users to construct a request and then either get the requests
+// response via Do(), parse the response into a simplejson.Json via JSON(),
+// or to parse the json response into an object via Into().
+type Builder interface {
+	WithContext(context.Context) Builder
+	WithBody(io.Reader) Builder
+	WithMethod(string) Builder
+	WithHeader(http.Header) Builder
+	SetHeader(key, value string) Builder
+	Do() (*http.Response, error)
+	Into(interface{}) error
+	JSON() (*simplejson.Json, error)
+}
+
+type builder struct {
+	context  context.Context
+	method   string
+	endpoint string
+	body     io.Reader
+	header   http.Header
+	response *http.Response
+}
+
+// New provides a new Builder for the given endpoint.
+func New(endpoint string) Builder {
+	return &builder{
+		endpoint: endpoint,
+		method:   "GET",
+	}
+}
+
+// WithContext adds a context to the request.
+// If no context is provided, context.Background() is used instead.
+func (r *builder) WithContext(ctx context.Context) Builder {
+	r.context = ctx
+	return r
+}
+
+// WithBody adds a body to the request.
+func (r *builder) WithBody(body io.Reader) Builder {
+	r.body = body
+	return r
+}
+
+// WithMethod sets the request method. Defaults to "GET".
+func (r *builder) WithMethod(method string) Builder {
+	r.method = method
+	return r
+}
+
+// WithHeader replaces the request header map with the given header map.
+func (r *builder) WithHeader(header http.Header) Builder {
```

Comment:
Another style preference question: I see the alignment to the singular naming of multiple headers in `http.Header` -- but the distinction of this setting all headers via a map vs the `SetHeader` function immediately below this setting a single header made it harder to tell `WithHeader` is intended to set all headers in bulk.

`WithHeaders` might be clearer for the API contract you are building, but then it deviates from the underlying source conventions.

Reply:
What about `WithHeaderMap`? Would that make sense?